pkg/controller/discovery/web/plan.go

```diff
@@ -0,0 +1,257 @@
+	"github.com/gin-gonic/gin"
+	auth "k8s.io/api/authorization/v1"
+	"k8s.io/api/core/v1"
+	"net/http"
+	"strings"
+)
+
+// Plan route root.
+const (
+	PlansRoot = Root + "/plans"
+	PlanRoot  = PlansRoot + "/:plan"
+)
+
+//
+// Pods request handler.
+type PlanHandler struct {
+	BaseHandler
+	plan model.Plan
+}
+
+func (h *PlanHandler) AddRoutes(r *gin.Engine) {
+	r.GET(PlansRoot, h.List)
+	r.GET(PlansRoot+"/", h.List)
+	r.GET(PlanRoot, h.Get)
+	r.GET(PlanRoot+"/pods", h.Pods)
+}
+
+func (h *PlanHandler) Prepare(ctx *gin.Context) int {
+	status := h.BaseHandler.Prepare(ctx)
+	if status != http.StatusOK {
+		return status
+	}
+	h.plan = model.Plan{
+		Base: model.Base{
+			Namespace: h.ctx.Param("namespace"),
+			Name:      h.ctx.Param("plan"),
+		},
+	}
+	err := h.plan.Select(h.container.Db)
+	if err != nil {
+		Log.Trace(err)
+		return http.StatusInternalServerError
+	}
+	status = h.allow(h.getSAR())
+	if status != http.StatusOK {
+		return status
+	}
+
+	return http.StatusOK
+}
+
+func (h *PlanHandler) getSAR() auth.SelfSubjectAccessReview {
+	return auth.SelfSubjectAccessReview{
+		Spec: auth.SelfSubjectAccessReviewSpec{
+			ResourceAttributes: &auth.ResourceAttributes{
+				Group:     "apps",
+				Resource:  "MigPlan",
+				Namespace: h.plan.Namespace,
+				Name:      h.plan.Name,
+				Verb:      "get",
+			},
+		},
+	}
+}
+
+func (h *PlanHandler) Get(ctx *gin.Context) {
+	status := h.Prepare(ctx)
+	if status != http.StatusOK {
+		h.ctx.Status(status)
+		return
+	}
+	err := h.plan.Select(h.container.Db)
+	if err != nil {
+		Log.Trace(err)
+		h.ctx.Status(http.StatusInternalServerError)
+		return
+	}
+	if !h.plan.HasPK() {
+		h.ctx.Status(http.StatusNotFound)
+		return
+	}
+	content := Plan{
+		Namespace:   h.plan.Namespace,
+		Name:        h.plan.Name,
+		Source:      h.plan.DecodeSource(),
+		Destination: h.plan.DecodeDestination(),
+	}
+
+	h.ctx.JSON(http.StatusOK, content)
+}
+
+func (h *PlanHandler) Pods(ctx *gin.Context) {
+	status := h.Prepare(ctx)
+	if status != http.StatusOK {
+		h.ctx.Status(status)
+		return
+	}
+	content := PlanPods{}
+	content.With(h)
+
+	h.ctx.JSON(http.StatusOK, content)
+}
+
+func (h *PlanHandler) List(ctx *gin.Context) {
+	status := h.Prepare(ctx)
+	if status != http.StatusOK {
+		h.ctx.Status(status)
+		return
+	}
+	list, err := model.PlanList(h.container.Db, &h.page)
+	if err != nil {
+		Log.Trace(err)
+		h.ctx.Status(http.StatusInternalServerError)
+		return
+	}
+	content := []Plan{}
+	for _, m := range list {
+		d := Plan{
+			Namespace:   m.Namespace,
+			Name:        m.Name,
+			Source:      m.DecodeSource(),
+			Destination: m.DecodeDestination(),
+		}
+		content = append(content, d)
+	}
+
+	h.ctx.JSON(http.StatusOK, content)
+}
+
+//
+// Plan REST resoruce.
+type Plan struct {
+	Namespace   string
+	Name        string
+	Source      *v1.ObjectReference
+	Destination *v1.ObjectReference
+}
+
+//
+// PlanPods REST resource.
+type PlanPods struct {
+	Namespace   string
+	Name        string
+	Controller  Pod
```

Comment:
Would be simpler to have it as a list of pods here too.

Reply:
How so?  There should only be 1 controller pod.  Right?

Reply:
Just in terms of generic implementation of the UI client, it is simpler to treat it as a list of log sources. But you are right, for now it is 1 controller pod, unless something will change (re: https://docs.google.com/document/d/1632Iw2DCnpMe7zMhL3YOmDPI9EPuEA5474hgTafxqrc/edit?disco=AAAADjiS2bE)

Reply:
Easy enough. will do.

Reply:
Done.